shared/dataflow/codeql/dataflow/VariableCapture.qll

```diff
@@ -0,0 +1,930 @@
+  class ClosureExpr extends Expr {
+    /**
+     * Holds if `body` is the callable body of this closure. A lambda expression
+     * only has one body, but in general a locally defined object may have
+     * multiple such methods and constructors.
+     */
+    predicate hasBody(Callable body);
+
+    /**
+     * Holds if `f` is an expression that may hold the value of the closure and
+     * may occur in a position where the value escapes or where the closure may
+     * be invoked.
+     *
+     * For example, if a lambda is assigned to a variable, then references to
+     * that variable in return or argument positions should be included.
+     */
+    predicate hasAliasedAccess(Expr f);
+  }
+
+  class Callable {
+    /** Gets a textual representation of this callable. */
+    string toString();
+
+    /** Gets the location of this callable. */
+    Location getLocation();
+
+    /** Holds if this callable is a constructor. */
+    predicate isConstructor();
+  }
+}
+
+signature module OutputSig<InputSig I> {
+  /**
+   * A data flow node that we need to reference in the step relations for
+   * captured variables.
+   */
+  class ClosureNode;
+
+  /**
+   * A synthesized data flow node representing the storage of a captured
+   * variable.
+   */
+  class SynthesizedCaptureNode extends ClosureNode {
+    /** Gets a textual representation of this node. */
+    string toString();
+
+    /** Gets the location of this node. */
+    I::Location getLocation();
+
+    /** Gets the enclosing callable. */
+    I::Callable getEnclosingCallable();
+
+    /** Holds if this node is a synthesized access of `v`. */
+    predicate isVariableAccess(I::CapturedVariable v);
+
+    /** Holds if this node is a synthesized instance access. */
+    predicate isInstanceAccess();
+  }
+
+  /** A data flow node for an expression. */
+  class ExprNode extends ClosureNode {
+    /** Gets the expression corresponding to this node. */
+    I::Expr getExpr();
+  }
+
+  /** A data flow node for the `PostUpdateNode` of an expression. */
+  class ExprPostUpdateNode extends ClosureNode {
+    /** Gets the expression corresponding to this node. */
+    I::Expr getExpr();
+  }
+
+  /** A data flow node for a parameter. */
+  class ParameterNode extends ClosureNode {
+    /** Gets the parameter corresponding to this node. */
+    I::CapturedParameter getParameter();
+  }
+
+  /** A data flow node for an instance parameter. */
+  class ThisParameterNode extends ClosureNode {
+    /** Gets the callable this instance parameter belongs to. */
+    I::Callable getCallable();
+  }
+
+  /** A data flow node for the instance parameter argument of a constructor call. */
+  class MallocNode extends ClosureNode {
+    /** Gets the closure construction that is the post-update of this node. */
+    I::ClosureExpr getClosureExpr();
+  }
+
+  /** Holds if `post` is a `PostUpdateNode` for `pre`. */
+  predicate capturePostUpdateNode(SynthesizedCaptureNode post, SynthesizedCaptureNode pre);
```

Comment:
Obviously one would include qldoc to clarify, but I think it would still potentially signal that they are always post-update nodes, which of course isn't the case. Then it could be something like `isPostUpdate(SynthesizedCaptureNode pre)`, but I'm not sure that's better either.
As-is this currently mirrors `predicate summaryPostUpdateNode(SummaryNode post, SummaryNode pre)`.